packages/core/content-releases/server/src/services/release.ts

```diff
@@ -460,166 +460,192 @@ const createReleaseService = ({ strapi }: { strapi: LoadedStrapi }) => ({
+            action: action.type,
+            id: action.entry.id,
+          });
         }
       }
 
-      // Then, we can continue with publishing the collectionTypes
-      for (const contentTypeUid of Object.keys(collectionTypeActions)) {
-        // @ts-expect-error - populateBuilderService should be a function but is returning service
-        const populate = await populateBuilderService(contentTypeUid)
-          .populateDeep(Infinity)
-          .build();
-
-        const { entriestoPublishIds, entriesToUnpublishIds } =
-          collectionTypeActions[contentTypeUid as UID.ContentType];
-
-        /**
-         * We need to get the populate entries to be able to publish without errors on components/relations/dynamicZones
-         * Considering that populate doesn't work well with morph relations we can't get the entries from the Release model
-         * So, we need to fetch them manually
-         */
-        const entriesToPublish = (await strapi.entityService.findMany(
-          contentTypeUid as UID.ContentType,
-          {
-            filters: {
-              id: {
-                $in: entriestoPublishIds,
-              },
-            },
-            populate,
+      const entityManagerService = strapi.plugin('content-manager').service('entity-manager');
+      const populateBuilderService = strapi.plugin('content-manager').service('populate-builder');
+
+      // Only publish the release if all action updates are applied successfully to their entry, otherwise leave everything as is
+      await strapi.db.transaction(async () => {
+        // First we publish all the singleTypes
+        for (const { uid, action, id } of singleTypeActions) {
+          // @ts-expect-error - populateBuilderService should be a function but is returning service
+          const populate = await populateBuilderService(uid).populateDeep(Infinity).build();
+
+          const entry = await strapi.entityService.findOne(uid, id, { populate });
+
+          try {
+            if (action === 'publish') {
+              await entityManagerService.publish(entry, uid);
+            } else {
+              await entityManagerService.unpublish(entry, uid);
+            }
+          } catch (error) {
+            if (
+              error instanceof errors.ApplicationError &&
+              (error.message === 'already.published' || error.message === 'already.draft')
+            ) {
+              // We don't want throw an error if the entry is already published or draft
+            } else {
+              throw error;
+            }
           }
-        )) as Entity[];
+        }
 
-        const entriesToUnpublish = (await strapi.entityService.findMany(
-          contentTypeUid as UID.ContentType,
-          {
-            filters: {
-              id: {
-                $in: entriesToUnpublishIds,
+        // Then, we can continue with publishing the collectionTypes
+        for (const contentTypeUid of Object.keys(collectionTypeActions)) {
```

Comment:
Not for this PR but idea for refactor:

Same thing maybe we can continue to clean up this massive function by organizing these into two more functions:

`publishSingleTypes` and `publishCollectionTypes`